rust/lance-datafusion/src/substrait.rs

```diff
@@ -271,22 +271,31 @@ pub async fn parse_substrait(expr: &[u8], input_schema: Arc<Schema>) -> Result<E
         }),
     }?;
 
-    let (substrait_schema, input_schema, index_mapping) =
-        remove_extension_types(envelope.base_schema.as_ref().unwrap(), input_schema.clone())?;
+    let (substrait_schema, input_schema) =
+        if envelope.base_schema.as_ref().unwrap().r#struct.is_some() {
```

Comment:
The zone maps / substrait / datafusion stuff is a bit of a hack atm.  I am trying really hard to avoid datafusion being a dependency of the file reader.  Unfortunately, that means we need _something_ to represent a filter expression in the file reader and so I'm using substrait.

However, lance (top level / dataset package) IS using datafusion for its filters.  So this means we go from dataset filter (datafusion) to substrait and then (in the decoder plugin) back to a datafusion filter.

Ideally there would just be a lightweight expressions library I could use and, once I figure out cloning, it might be fun to build one.

In the meantime this hack is because we need to encode DF filters into substrait somehow and, when I do so (in `encode_substrait` above) I've being lazy and not including the `r#struct` field which we're using here to detect if extension types are used.

Reply:
> Ideally there would just be a lightweight expressions library I could use

One could argue that DataFusion's goal is to be modular like that, and `datafusion-expr` could serve this purpose. However, I think the current list of dependencies make it not so lightweight: https://crates.io/crates/datafusion-expr/42.0.0/dependencies

Reply:
Hmm, if we can make sqlparser optional and get rid of arrow (arrow-buffer and arrow-array are fine) then it would probably be fine (I don't know how big those datafusion-...-common crates are but hopefully they're pretty small).  I'd also want to drag the substrait expression parsing into the crate (under a feature flag probably).  I'll add it to my potential list of xmas break projects :laughing: